src/login/mod.rs

```diff
@@ -1,104 +1,220 @@
-        .lines()
-        .filter(|line| !line.starts_with("---"))
-        .fold(String::new(), |mut data, line| {
-            data.push_str(line);
-            data
-        });
-    let pubkey_encoded = percent_encode(pubkey_filtered.as_bytes(), NON_ALPHANUMERIC).to_string();
+use crate::cli::login::SCOPES_LIST;
+use crate::commands::config::global_config;
+use crate::commands::logout::revoke_token;
+use crate::login::http::http_server_get_params;
+use crate::settings::{get_global_config_path, global_user::GlobalUser};
+
+pub static CLIENT_ID: &str = "54d11594-84e4-41aa-b438-e81b8fa78ee7";
+pub static AUTH_URL: &str = "https://dash.cloudflare.com/oauth2/auth";
+static TOKEN_URL: &str = "https://dash.cloudflare.com/oauth2/token";
+static CALLBACK_URL: &str = "http://localhost:8976/oauth/callback";
+
+pub fn run(scopes: Option<&[String]>) -> Result<()> {
+    // Create oauth2 client
+    let client = BasicClient::new(
+        ClientId::new(CLIENT_ID.to_string()),
+        None,
+        AuthUrl::new(AUTH_URL.to_string()).expect("Invalid authorization endpoint URL"),
+        Some(TokenUrl::new(TOKEN_URL.to_string()).expect("Invalid token endpoint URL")),
+    )
+    .set_redirect_uri(RedirectUrl::new(CALLBACK_URL.to_string()).expect("Invalid redirect URL"))
+    .set_auth_type(AuthType::RequestBody);
+
+    let (pkce_challenge, pkce_verifier) = PkceCodeChallenge::new_random_sha256();
+
+    // Create URL for user with the necessary scopes
+    let mut client_state = client
+        .authorize_url(CsrfToken::new_random)
+        .set_pkce_challenge(pkce_challenge);
+
+    if scopes.is_none() {
+        // User did not provide any scopes
+        for scope in SCOPES_LIST {
+            client_state = client_state.add_scope(Scope::new(scope.to_string()));
+        }
+    } else {
+        // User did provide some scopes
+        for scope in scopes.unwrap() {
+            client_state = client_state.add_scope(Scope::new(scope.to_string()));
+        }
+    }
+    client_state = client_state.add_scope(Scope::new("offline_access".to_string()));
+    let (auth_url, csrf_state) = client_state.url();
 
+    // Navigate to authorization endpoint
     let browser_permission =
         interactive::confirm("Allow Wrangler to open a page in your browser?")?;
     if !browser_permission {
         anyhow::bail!("In order to log in you must allow Wrangler to open your browser. If you don't want to do this consider using `wrangler config`");
     }
 
-    open_browser(&format!(
-        "https://dash.cloudflare.com/wrangler?key={0}",
-        pubkey_encoded
-    ))?;
-
-    let encrypted_token_str = poll_token(pubkey_filtered)?;
-    let encrypted_token = base64::decode_block(encrypted_token_str.as_str())?;
-    let mut token_bytes: [u8; 128] = [0; 128];
-
-    rsa.private_decrypt(&encrypted_token, &mut token_bytes, Padding::PKCS1)?;
-    let token = str::from_utf8(&token_bytes)?.trim_matches(char::from(0));
+    if wsl::is_wsl() {
```

Comment:
Is there a better way to check if the current process is run on WSL without having to include another dependency?